src/Physics/NuElectron/XSection/PXSecOnElectron.cxx

```diff
@@ -0,0 +1,126 @@
+#include "Physics/NuElectron/XSection/PXSecOnElectron.h"
+#include "Framework/Messenger/Messenger.h"
+#include "Framework/ParticleData/PDGUtils.h"
+#include "Framework/Utils/KineUtils.h"
+#include "Physics/NuclearState/ElectronVelocity.h"
+#include "Framework/Numerical/MathUtils.h"
+
+#include "TFile.h"
+#include "TGraph.h"
+#include <fstream>
+#include <iterator>
+
+using namespace genie;
+using namespace genie::constants;
+using namespace genie::controls;
+
+//____________________________________________________________________________
+PXSecOnElectron::PXSecOnElectron(string name, string config) :
+XSecAlgorithmI(name, config)
+{
+
+}
+//____________________________________________________________________________
+PXSecOnElectron::PXSecOnElectron() :
+XSecAlgorithmI("genie::PXSecOnElectron")
+{
+
+}
+//____________________________________________________________________________
+PXSecOnElectron::~PXSecOnElectron()
+{
+
+}
+//____________________________________________________________________________
+double PXSecOnElectron::Integral(const Interaction * interaction) const
+{
+  double xsec_sum = 0; 
+  double xsec_sum2 = 0; 
+  int NInt = 0; //Count number of integrations
+  do{
+    NInt++;
+    Interaction in_curr(*interaction); //Copy interaction object
+    fElectronVelocity->InitializeVelocity(in_curr); //Modify interaction to give electron random velocity from selected distribution
+    double xsec = fXSecIntegrator->Integrate(this,&in_curr); // In ele at rest
+    
+    //get gamma - comps orthogonal to beam x,y
+    //scale = sqrt(1-b_t^2)
+    auto probe_direction = in_curr.InitState().GetProbeP4(kRfHitElRest)->Vect().Unit();
+    auto electron_mom = in_curr.InitState().Tgt().HitPartP4();
+    auto transverse_mom = utils::math::GetOrthogonal(electron_mom, probe_direction);
+    auto gamma_transverse = transverse_mom.Gamma();
+
+    xsec *= gamma_transverse; //Correct for lorentz factor
+    xsec_sum+=xsec;
+    xsec_sum2+=TMath::Power(xsec,2);
+    double xsec_mean = xsec_sum/NInt;
+    //var = (sum(xi^2)/N-xsec_mean^2)
+    //rel_err = sigma/sqrt(n)*mean
+    double xsec_err = sqrt((xsec_sum2/NInt-TMath::Power(xsec_mean,2))/NInt)/xsec_mean;
+    if (NInt > 1 && xsec_err < fErrTolerance){break;} //Break condition for dipping below set tolerance
+  }
+  while ( NInt < fNIntegration); 
+  double xsec_avg = xsec_sum/NInt;
+  return xsec_avg;
+}
+//____________________________________________________________________________
+bool PXSecOnElectron::ValidProcess(const Interaction * interaction) const
+{
+  bool is_valid = true;
+  if(!pdg::IsElectron(interaction->InitState().TgtPtr()->HitPartPdg())) is_valid = false;
+  if(!interaction->TestBit(kISkipProcessChk)) is_valid = false;
+  return is_valid;
+}
+//____________________________________________________________________________
+bool PXSecOnElectron::ValidKinematics(const Interaction* interaction) const
+{
+  if(interaction->TestBit(kISkipKinematicChk)) return true;
```

Comment:
We still need to check the energy threshold. I think here should be right place. unless it's done somewhere else.